osf/models/preprint_service.py

```diff
@@ -260,3 +323,729 @@ def _send_preprint_confirmation(self, auth):
+        contributor.bibliographic = contributor_obj.visible
+        contributor.node_id = self._id
+        contributor_order = list(self.get_preprintcontributor_order())
+        contributor.index = contributor_order.index(contributor_obj.pk)
+
+        if save:
+            contributor.save()
+
+        return contributor_obj
+
+    def set_visible(self, user, visible, log=True, auth=None, save=False):
+        if not self.is_contributor(user):
+            raise ValueError(u'User {0} not in contributors'.format(user))
+
+        if visible and not PreprintContributor.objects.filter(preprint=self, user=user, visible=True).exists():
+            PreprintContributor.objects.filter(preprint=self, user=user, visible=False).update(visible=True)
+        elif not visible and PreprintContributor.objects.filter(preprintservice=self, user=user, visible=True).exists():
+            if PreprintContributor.objects.filter(preprint=self, visible=True).count() == 1:
+                raise ValueError('Must have at least one visible contributor')
+            PreprintContributor.objects.filter(preprint=self, user=user, visible=True).update(visible=False)
+        else:
+            return
+        # todo: implement preprint log
+        # message = (
+        # NodeLog.MADE_CONTRIBUTOR_VISIBLE
+        # if visible
+        # else NodeLog.MADE_CONTRIBUTOR_INVISIBLE
+        # )
+        # if log:
+        # self.add_log(
+        # message,
+        # params={
+        # 'parent': self.parent_id,
+        # 'node': self._id,
+        # 'contributors': [user._id],
+        # },
+        # auth=auth,
+        # save=False,
+        # )
+        if save:
+            self.save()
+
+    def replace_contributor(self, old, new):
+        try:
+            contrib_obj = self.preprintcontributor_set.get(user=old)
+        except PreprintContributor.DoesNotExist:
+            return False
+        contrib_obj.user = new
+        contrib_obj.save()
+        for group_name in self.group_names:
+            if old in self.get_group(group_name).user_set:
+                self.get_group(group_name).user_set.remove(old)
+                self.get_group(group_name).user_set.add(new)
+
+        # Remove unclaimed record for the project
+        if self._id in old.unclaimed_records:
+            del old.unclaimed_records[self._id]
+            old.save()
+        # self.save_node_preprints()  # TODO: decide what to do with this
+        return True
+
+    def remove_contributor(self, contributor, auth, log=True):
+        """Remove a contributor from this node.
+
+        :param contributor: User object, the contributor to be removed
+        :param auth: All the auth information including user, API key.
+        """
+
+        if isinstance(contributor, PreprintContributor):
+            contributor = contributor.user
+
+        # remove unclaimed record if necessary
+        if self._primary_key in contributor.unclaimed_records:
```

Comment:
Minor: `_primary_key` is deprecated, since it's just there for backwards-compat. Use `._id` instead.